File: github_cpp/18/64.cpp
#include <iostream>
#include <cstring>
#include <list>

using namespace std;
int price[] = {1, 5, 8, 9, 10, 17, 17, 20, 24, 30};
int rodCutting(int rod[], int length);
int bottom_up(int rod[], int n);
void get_pieces(int s[], int n);
int rod_cut(int rod[],int length);
int r[10];
int max(int a, int b);
int length = 10;
int main()
{
    memset(r, 0, sizeof(int)*10);
    cout<<"revenue: "<<rodCutting(price, 4);

    return 0;
}

int rod_cut(int rod[], int length)
{
    if(length == 0) return 0;
    int maximum = -1;
    for(int i=1;i<=length;++i)
    {
        maximum = max(maximum,rod[i]+rod_cut(rod,length-1));
    }
    return maximum;
}


int rodCutting(int rod[], int length)
{
    if(length == 1) return rod[0];
    if(r[length-1] > 0) return r[length-1];
    int maximum = -1;
    for(int i = 1; i < length; ++i)
    {
        int curr = max(rod[length-1], rod[i-1] + rodCutting(rod, length-i));
        maximum = max(maximum, curr);
    }
    r[length-1] = maximum;
  return maximum;
}
int bottom_up(int rod[], int n)
{
    int r[n+1];
    int s[n+1];
    r[0] = 0;
    for(int i = 1; i <= n; ++i)
    {
        r[i] = rod[i-1];
    }
    for(int i = 1; i <= n; ++i)
    {
        int maximum = -1;
        for(int j = 1; j <= i; ++j)
        {
            if(maximum < r[j] + r[i-j])
            {
                maximum = r[j] + r[i-j];
                s[i] = j;
            }
        }
        r[i] = maximum;
    }
    get_pieces(s, n);
    return r[n];
}
void get_pieces(int s[], int n)
{
    cout<<"Using pieces:\n";
    while(n > 0)
    {
        cout<<s[n]<<endl;
        n = n - s[n];
    }
}
int max(int a, int b)
{
    return (a > b ? a : b);
}

File: github_code_index/java/15/64.java
package com.anuragkapur.misc;

import java.util.HashMap;
import java.util.Map;

/**
 * Problem Statement:
 *
 * Given a rod of length n inches and a table of prices pi for i = 1,2,...,n, determine the maximum revenue rn
 * obtainable by cutting up the rod and selling the pieces. Note that if the price pn for a rod of length n is large
 * enough, an optimal solution may require no cutting at all.
 *
 * @author anuragkapur
 */
public class RodCutting {

    /**
     * Simple recursive approach to find the max price
     *
     * Run time complexity: O(2^n) ; tries all possible cut permutations
     *
     * @param p
     * @param n
     * @return
     */
    public int cutRodToMaximisePrice1(int p[], int n) {

        if (n == 0) {
            return 0;
        }

        // a simple sysout to show that sub-problems are re-evaluated and hence there is scope for improvement in this
        // algorithm.
        System.out.println("n="+n);

        int maxPrice = Integer.MIN_VALUE;
        for (int i=0; i<n; i++) {
            maxPrice = Math.max(maxPrice, p[i]+cutRodToMaximisePrice1(p, n-i-1));
        }

        return maxPrice;
    }

    /**
     * Top down with memoization
     *
     * Run time complexity: O(n^2)
     *
     * @param p
     * @param n
     * @return
     */
    public int cutRodToMaximisePrice2(int p[], int n, Map<Integer, Integer> memoizedResults) {

        // recursion base case
        if (n == 0) {
            return 0;
        }

        // one time init for memoization
        if (memoizedResults == null) {
            System.out.println("Memoized results store init");
            memoizedResults = new HashMap<>();
        }

        if (memoizedResults.containsKey(n)) {
            return memoizedResults.get(n);
        }

        System.out.println("n="+n);

        int maxPrice = Integer.MIN_VALUE;
        for (int i=0; i<n; i++) {
            maxPrice = Math.max(maxPrice, p[i]+cutRodToMaximisePrice2(p, n-i-1, memoizedResults));
        }

        memoizedResults.put(n, maxPrice);

        return maxPrice;
    }
}
